smart_contract/src/lib.rs

```diff
@@ -264,6 +264,58 @@ where
     }
 }
 
+/// Defines an object that is part of the [`ExecutorDataModel`].
+///
+/// An object consists of id and payload.
+/// ID is a unique type identifier in the executor's data model.
+/// Payload is JSON-serialized type structure, according to the type schema.
+// TODO: support both JSON and SCALE for payload.
+pub trait ExecutorDataModelObject: Sized {
```

Comment:
I don't think there is much point in having this trait just for 2 types. I'd rather if we had 2 traits - `Permission` and `Parameter`

Reply:
Well, yeah, it seems to bring more boilerplate than benefit.

I wanted to formalize that "data model objects are those which have id and payload" and base impls on that. But, it turns out I can remove a lot and just leave that duplicated fragment for both `Parameter` and `Permission` traits in `iroha_executor`:

```rust
/// Try to convert from [`ParameterObject`]
/// # Errors
/// See [`TryFromDataModelObjectError`]
fn try_from_object(
    object: &ParameterObject,
) -> crate::prelude::Result<Self, TryFromDataModelObjectError> {
    if &object.id() != Self::id() {
        return Err(TryFromDataModelObjectError::Id(object.id().name().clone()));
    }
    object
        .payload()
        .deserialize()
        .map_err(TryFromDataModelObjectError::Deserialize)
}

/// Convert into [`ParameterObject`]
fn to_object(&self) -> ParameterObject {
    ParameterObject::new(
        Self::id(),
        JsonString::serialize(&self)
            .expect("failed to serialize concrete data model entity; this is a bug"),
    )
}

```